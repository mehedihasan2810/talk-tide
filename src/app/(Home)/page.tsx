import BgGradients from "@/components/Home/BgGradients";
import Hero from "@/components/Home/Hero";
import MockMessageItems from "@/components/Home/MockMessageItems";

export default async function Home() {
  return (
    <section className="relative isolate grid h-screen w-screen place-items-center overflow-hidden">
      <Hero />

      <BgGradients />

      <MockMessageItems />
    </section>
  );
}
