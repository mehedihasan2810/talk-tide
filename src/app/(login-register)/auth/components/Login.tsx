"use client";

import { zodResolver } from "@hookform/resolvers/zod";
import { useForm } from "react-hook-form";
import * as z from "zod";
import { Button } from "@/components/ui/button";
import { Alert, AlertDescription, AlertTitle } from "@/components/ui/alert";

import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormMessage,
} from "@/components/ui/form";
import { Input } from "@/components/ui/input";
import {
  ArrowPathIcon,
  ExclamationTriangleIcon,
  LockClosedIcon,
  XMarkIcon,
} from "@heroicons/react/20/solid";
import { signIn, useSession } from "next-auth/react";
import { LoginSchema } from "@/utils/zod-schema/loginSchema";
import { useState } from "react";
// import Link from "next/link";
// import { Separator } from "@/components/ui/separator";
import { useRouter } from "next/navigation";
import { useToast } from "@/components/ui/use-toast";

interface Props {
  updateIsLogin(): void;
}

export default function Login({ updateIsLogin }: Props) {
  const [error, setError] = useState<string | null | undefined>(null);
  const [isLoading, setIsLoading] = useState<boolean>(false);
  // const [isLoginSuccess, setIsLoginSuccess] = useState(false);

  const session = useSession();
  const router = useRouter();
  const { toast } = useToast();

  // useForm
  const form = useForm<z.infer<typeof LoginSchema>>({
    resolver: zodResolver(LoginSchema),
    defaultValues: {
      username: "",
      password: "",
    },
  });

  async function onSubmit(values: z.infer<typeof LoginSchema>) {
    // if user is already logged in then don't proceed. return with a message
    if (session.data) {
      setError("You are already logged in");
      return;
    }

    setIsLoading(true); // track log in loading states

    // log in the user with required credentials
    const res = await signIn("credentials", {
      redirect: false,
      authType: "login",
      ...values,
    });
    setIsLoading(false);

    // if login is not successful then show a error message
    if (!res?.ok || res.error) {
      setError(res?.error);
    } else {
      form.reset();
      toast({ title: "Logged in successfully", variant: "success" });
      router.refresh();
      router.replace("/chat");
      // setIsLoginSuccess(true);
    }
  }

  return (
    <div className="grid h-screen w-screen place-items-center">
      <Form {...form}>
        <form
          data-testid="login-form"
          onSubmit={form.handleSubmit(onSubmit)}
          className="w-80 space-y-5"
        >
          {/* header starts */}
          <div className="mb-8 flex items-center justify-center gap-1 text-primary">
            <h2 className="scroll-m-20 text-3xl font-semibold tracking-tight">
              Login
            </h2>
            <LockClosedIcon className="h-7 w-7" />
          </div>
          {/* header ends */}

          {/* Alert starts */}
          {error && (
            <Alert variant="destructive" className="relative">
              <button
                onClick={() => setError(null)}
                className="absolute right-2 top-1 rounded-full bg-rose-50 p-1"
                type="button"
              >
                <XMarkIcon className="h-5 w-5" />
              </button>
              <ExclamationTriangleIcon className="h-4 w-4" />
              <AlertTitle>Error!</AlertTitle>
              <AlertDescription>{error}</AlertDescription>
            </Alert>
          )}
          {/* Alert ends */}

          <FormField
            control={form.control}
            name="username"
            render={({ field }) => (
              <FormItem>
                <FormControl>
                  <Input
                    data-testid="login-username"
                    placeholder="Enter your username..."
                    {...field}
                    className="h-11 border-primary/40 text-base focus-visible:ring-primary"
                  />
                </FormControl>

                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="password"
            render={({ field }) => (
              <FormItem>
                <FormControl>
                  <Input
                    data-testid="login-password"
                    type="password"
                    placeholder="Enter your password..."
                    {...field}
                    className="h-11 border-primary/40 text-base focus-visible:ring-primary"
                  />
                </FormControl>

                <FormMessage />
                <FormDescription className="text-center">
                  Don&#39;t have an account?{" "}
                  <button
                    type="button"
                    disabled={isLoading}
                    onClick={updateIsLogin}
                    data-testid="register-link"
                    className="text-primary underline hover:text-primary/70"
                  >
                    Register
                  </button>
                </FormDescription>
              </FormItem>
            )}
          />
          <Button
            disabled={isLoading || session.status === "loading"}
            type="submit"
            className="h-11 w-full text-base"
          >
            {isLoading && (
              <ArrowPathIcon className="mr-2 h-4 w-4 animate-spin" />
            )}{" "}
            Login
          </Button>

          {/* Login success alert start */}

          {/* {isLoginSuccess && (
            <Alert
              variant="default"
              className="relative border border-teal-400 text-teal-500"
            >
              <button
                onClick={() => setIsLoginSuccess(false)}
                className="absolute right-2 top-1 rounded-full bg-teal-50 p-1 hover:opacity-50"
                type="button"
              >
                <XMarkIcon className="h-5 w-5" />
              </button>
              <ExclamationTriangleIcon className="h-4 w-4" />
              <AlertTitle>Logged in successfully!</AlertTitle>
              <AlertDescription>
                <div>You have logged in successfully</div>
                <div className="flex h-4 gap-2">
                  <div>
                    Go{" "}
                    <Link
                      href="/"
                      className="font-semibold underline hover:opacity-50"
                    >
                      Home
                    </Link>
                  </div>
                  <Separator color="teal" orientation="vertical" />
                  <div>
                    Go to{" "}
                    <Link
                      href="/chat"
                      className="font-semibold underline hover:opacity-50"
                    >
                      My Inbox
                    </Link>
                  </div>
                </div>
              </AlertDescription>
            </Alert>
          )} */}

          {/* Login success alert start */}
        </form>
      </Form>
    </div>
  );
}
