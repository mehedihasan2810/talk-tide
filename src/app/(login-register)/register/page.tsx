"use client";

import { zodResolver } from "@hookform/resolvers/zod";
import { useForm } from "react-hook-form";
import * as z from "zod";
import { Button } from "@/components/ui/button";
import {
  Form,
  FormControl,
  FormDescription,
  FormField,
  FormItem,
  FormMessage,
} from "@/components/ui/form";
import { Input } from "@/components/ui/input";
import Link from "next/link";
import { LockClosedIcon } from "@heroicons/react/20/solid";
import { signIn, useSession } from "next-auth/react";
import { RegisterSchema } from "@/utils/zod-schema/registerSchema";

interface Props {
  updateIsLogin(): void;
}

export default function Register({ updateIsLogin }: Props) {
  const session = useSession();
  console.log(session);

  // useForm starts ------------
  const form = useForm<z.infer<typeof RegisterSchema>>({
    resolver: zodResolver(RegisterSchema),
    defaultValues: {
      email: "",
      username: "",
      password: "",
    },
  });
  // useForm ends ----------

  async function onSubmit(values: z.infer<typeof RegisterSchema>) {
    console.log(values);
    const data = await signIn("credentials", {
      redirect: false,
      // redirect: true,
      // callbackUrl: "/",
      authType: "register",
      ...values,
    });
    console.log(data);
  }
  return (
    <div className="w-screen h-screen grid place-items-center">
      <Form {...form}>
        <form
          data-testid="register-form"
          onSubmit={form.handleSubmit(onSubmit)}
          className="space-y-5 w-80"
        >
          {/* header starts */}
          <div className="flex justify-center items-center gap-1 text-emerald-700 mb-8">
            <h2 className="scroll-m-20 text-3xl font-semibold tracking-tight">
              Register
            </h2>
            <LockClosedIcon className="w-8 h-8" />
          </div>
          {/* header ends */}
          <FormField
            control={form.control}
            name="email"
            render={({ field }) => (
              <FormItem>
                <FormControl>
                  <Input
                    placeholder="Enter your email..."
                    {...field}
                    className="h-11 text-base focus-visible:ring-emerald-500 border-gray-300"
                  />
                </FormControl>
                <FormMessage />
              </FormItem>
            )}
          />

          <FormField
            control={form.control}
            name="username"
            render={({ field }) => (
              <FormItem>
                <FormControl>
                  <Input
                    placeholder="Enter your username..."
                    {...field}
                    className="h-11 text-base focus-visible:ring-emerald-500 border-gray-300"
                  />
                </FormControl>

                <FormMessage />
              </FormItem>
            )}
          />
          <FormField
            control={form.control}
            name="password"
            render={({ field }) => (
              <FormItem>
                <FormControl>
                  <Input
                    type="password"
                    placeholder="Enter your password..."
                    {...field}
                    className="h-11 text-base focus-visible:ring-emerald-500 border-gray-300"
                  />
                </FormControl>

                <FormMessage />
                <FormDescription className="text-center">
                  Already have an account?{" "}
                  <Link
                    onClick={updateIsLogin}
                    href="#"
                    className="text-emerald-500 hover:text-emerald-400 underline"
                  >
                    Login
                  </Link>
                </FormDescription>
              </FormItem>
            )}
          />
          <Button
            type="submit"
            className="w-full h-11 text-base bg-emerald-700 hover:bg-emerald-800"
          >
            Register
          </Button>
        </form>
      </Form>
    </div>
  );
}
