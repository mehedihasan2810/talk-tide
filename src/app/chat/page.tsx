"use client";
import React from "react";
import Sidebar from "./components/Sidebar";
import { useChatStore } from "@/lib/stores/chatStores";

const Chat = () => {
  const selectedUser = useChatStore((state) => state.selectedUser);
  const groupName = useChatStore((state) => state.groupName);
  const groupParticipants = useChatStore((state) => state.groupParticipants);
  console.log(groupParticipants);
  console.log(selectedUser);
  console.log(groupName);
  return (
    <div className="sm:hidden">
      <Sidebar />
    </div>
  );
};

export default Chat;
