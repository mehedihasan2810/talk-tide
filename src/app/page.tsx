export default function Home() {
  return (
    <div>
      <h1>helloooooo</h1>
    </div>
  );
}
