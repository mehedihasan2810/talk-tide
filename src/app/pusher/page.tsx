"use client";
import { Button } from "@/components/ui/button";
import { Input } from "@/components/ui/input";
import { apiClient } from "@/lib/api";
import { pusherClient } from "@/lib/pusher";
import { SessionUser } from "@/types/session";
import { useSession } from "next-auth/react";
import { useRouter } from "next/navigation";
import React, { useEffect, useState } from "react";

const Pusher = () => {
  const router = useRouter();

  const { data: session, status } = useSession({
    required: true,
    onUnauthenticated() {
      router.replace("/auth");
    },
  });
  const [msg, setMsg] = useState("");
  const [messages, setMessages] = useState<string[]>([]);

  const msgHandler = (msg: string) => {
    console.log(msg);
    setMessages((m) => [msg, ...m]);
  };

  useEffect(() => {
    if (status === "loading") return;
    const myEvent = pusherClient.subscribe((session.user as SessionUser).id);

    myEvent.bind("msg", msgHandler);

    return () => {
      pusherClient.unsubscribe((session.user as SessionUser).id);

      myEvent.unbind("msg", msgHandler);
    };
  }, [status]);

  return (
    <div className="fixed z-50 grid h-full w-full place-items-center">
      <div>
        {messages.map((m, i) => (
          <p key={i}>{m}</p>
        ))}
        <Input
          value={msg}
          onChange={(e) => {
            setMsg(e.target.value);
          }}
          type="text"
          placeholder="type a message"
        />
        <Button
          onClick={async () => {
            const res = await apiClient.post("/chatApp", { message: msg });
            console.log(res);
            setMsg("");
          }}
        >
          send
        </Button>
      </div>
    </div>
  );
};

export default Pusher;
