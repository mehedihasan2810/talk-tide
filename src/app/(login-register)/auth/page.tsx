"use client";
import React from "react";

import { usePathname, useRouter, useSearchParams } from "next/navigation";
import Login from "./components/Login";
import Register from "./components/Register";
import { useCreateQueryString } from "@/app/chat/hooks/useCreateQueryString";

const Auth = () => {
  const router = useRouter();
  const pathname = usePathname();
  const searchParams = useSearchParams()!;

  const createQueryString = useCreateQueryString();

  return (
    <section>
      {searchParams.get("authPage") !== "register" ? (
        <Login
          updateIsLogin={() => {
            // store authPage state in url so that we can
            // persist our page on page reloading
            router.replace(
              pathname + createQueryString("authPage", "register"),
            );
          }}
        />
      ) : (
        <Register
          updateIsLogin={() => {
            // store authPage state in url so that we can
            // persist our page on page reloading
            router.replace(pathname + createQueryString("authPage", "login"));
          }}
        />
      )}
    </section>
  );
};

Auth.pageName = "auth";

export default Auth;
