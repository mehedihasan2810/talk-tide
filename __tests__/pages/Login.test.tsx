import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Login from "@/app/(login-register)/login/page";
import { formSchema } from "@/app/(login-register)/login/page";
describe("<Login />", () => {
  it("should render, have Login title and match snapshot", () => {
    const { asFragment } = render(<Login />); // ARRANGE

    // ACT
    const loginTextEl = screen.getByRole("heading", {
      level: 2,
      name: "Login",
    });

    // ASSERT
    expect(loginTextEl).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  // -----------------------------------------------------------------

  it("Email, username, password input field should work", async () => {
    render(<Login />); //ARRANGE

    // ACT
    const user = userEvent.setup();
    const usernameInput = screen.getByPlaceholderText(
      "Enter your username..."
    ) as HTMLInputElement;
    const passwordInput = screen.getByPlaceholderText(
      "Enter your password..."
    ) as HTMLInputElement;

    // EVENT
    await user.type(usernameInput, "foo");
    await user.type(passwordInput, "foo");

    // ASSERT
    expect(usernameInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(passwordInput).toHaveAttribute("type", "password");
    expect(usernameInput.value).toEqual("foo");
    expect(passwordInput.value).toEqual("foo");
  });

  // -----------------------------------------------------------

  it("form data validation should work", async () => {
    render(<Login />); //ARRANGE

    // ACT
    const user = userEvent.setup();
    const formNode = screen.getByTestId("login-form") as HTMLFormElement;
    const submitBtn = screen.getByRole("button", { name: "Login" });
    const usernameInput = screen.getByTestId(
      "login-username"
    ) as HTMLInputElement;
    const passwordInput = screen.getByTestId(
      "login-password"
    ) as HTMLInputElement;

    // EVENT
    await user.type(usernameInput, "Mehedi Hasan");
    await user.type(passwordInput, "123456");

    const formData = Object.fromEntries(new FormData(formNode));
    // ASSERT
    expect(formSchema.parse(formData)).toStrictEqual(formData);
    expect(formNode).toBeInTheDocument();
    expect(submitBtn).toBeInTheDocument();
    expect(submitBtn).toHaveAttribute("type", "submit");
  });
});
