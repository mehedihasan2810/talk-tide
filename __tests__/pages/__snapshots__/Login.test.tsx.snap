// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render, have Login title and match snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-screen h-screen grid place-items-center"
  >
    <form
      class="space-y-5 w-80"
      data-testid="login-form"
    >
      <div
        class="flex justify-center items-center gap-1 text-emerald-700 mb-8"
      >
        <h2
          class="scroll-m-20 text-3xl font-semibold tracking-tight"
        >
          Login
        </h2>
        <svg
          aria-hidden="true"
          class="w-7 h-7"
          fill="currentColor"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10 1a4.5 4.5 0 00-4.5 4.5V9H5a2 2 0 00-2 2v6a2 2 0 002 2h10a2 2 0 002-2v-6a2 2 0 00-2-2h-.5V5.5A4.5 4.5 0 0010 1zm3 8V5.5a3 3 0 10-6 0V9h6z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
      <div
        class="space-y-2"
      >
        <input
          aria-describedby=":r0:-form-item-description"
          aria-invalid="false"
          class="flex w-full rounded-md border bg-background px-3 py-2 ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 h-11 text-base focus-visible:ring-emerald-500 border-gray-300"
          data-testid="login-username"
          id=":r0:-form-item"
          name="username"
          placeholder="Enter your username..."
          value=""
        />
      </div>
      <div
        class="space-y-2"
      >
        <input
          aria-describedby=":r1:-form-item-description"
          aria-invalid="false"
          class="flex w-full rounded-md border bg-background px-3 py-2 ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 h-11 text-base focus-visible:ring-emerald-500 border-gray-300"
          data-testid="login-password"
          id=":r1:-form-item"
          name="password"
          placeholder="Enter your password..."
          type="password"
          value=""
        />
        <p
          class="text-sm text-muted-foreground text-center"
          id=":r1:-form-item-description"
        >
          Don't have an account? 
          <a
            class="text-emerald-500 hover:text-emerald-400 underline"
            data-testid="login-link"
            href="register"
          >
            Register
          </a>
        </p>
      </div>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 text-primary-foreground px-4 py-2 w-full h-11 text-base bg-emerald-700 hover:bg-emerald-800"
        type="submit"
      >
        Login
      </button>
    </form>
  </div>
</DocumentFragment>
`;
